src/LaunchDarkly.EventSource/EventSource.cs

```diff
@@ -138,28 +144,62 @@ public async Task StartAsync()
                     }
                     await MaybeWaitWithBackOff();
                 }
-                firstTime = false;
+                firstTime = false;
+
+                CancellationTokenSource newRequestTokenSource = null;
+                CancellationToken cancellationToken;
+                lock (this)
+                {
+                    if (_readyState == ReadyState.Shutdown)
+                    {
+                        // in case Close() was called in between the previous ReadyState check and the creation of the new token
+                        return;
+                    }
+                    newRequestTokenSource = new CancellationTokenSource();
+                    _currentRequestToken?.Dispose();
+                    _currentRequestToken = newRequestTokenSource;
+                }
+
+                if (ReadyState == ReadyState.Connecting || ReadyState == ReadyState.Open)
+                {
+                    throw new InvalidOperationException(string.Format(Resources.EventSource_Already_Started, ReadyState));
+                }
+
+                SetReadyState(ReadyState.Connecting);
+                cancellationToken = newRequestTokenSource.Token;
+
```

Comment:
This logic is more or less equivalent to what we had, but some of it was down in the `ConnectToEventSourceAsync` method, and I pulled it up here because it wasn't that closely related to the main task of that method (setting up the `EventSourceService`). We may not even need that method— it's never used anywhere else.